### PR TITLE
[REVIEW] Refactor `detail::gather` API to use scoped enumerators [skip ci]

### PR DESCRIPTION
**Pull request by codereport**
In PR https://github.com/rapidsai/cudf/pull/5191 I noticed the back-to-back-to-back `bool`s in the following API:
```cpp
    grouped_values = cudf::experimental::detail::gather(
      values, helper().key_sort_order(), false, false, false, mr);
```
I [posted a comment](https://github.com/rapidsai/cudf/pull/5191#discussion_r425456672) about it to address it in a follow up PR. This is that PR.

Note that a couple of the calls to `detail::gather` were completely skipping the three `bool`s and only passing `mr` and `stream` (in the wrong order as well). Due to the fact that all of the bools had default values though and C++ wonderful implicit conversions to `bool`, this code compiled fine. 

**Please pay careful attention to said calls:**
* `strings/copying/copying.cu`
* `strings/sorting/sorting.cu`

PS - Note also that in `add_keys.cu`, this comment:
```
false,  // ignore out-of-bounds
```
was actually on the wrong line.

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

File: cpp/include/cudf/detail/gather.hpp [hunk 1]
```diff
@@ -41,9 +48,9 @@ namespace detail {
  */
 std::unique_ptr<table> gather(table_view const& source_table,
                               column_view const& gather_map,
-                              bool check_bounds                   = false,
-                              bool ignore_out_of_bounds           = false,
-                              bool allow_negative_indices         = false,
+                              bounds check_bounds                 = bounds::NO_CHECK,
+                              out_of_bounds oob                   = out_of_bounds::DONT_IGNORE,
```
Comment (vuule) [inline cpp/include/cudf/detail/gather.hpp:52]: check_bounds should not have check in the name, but I can't think of a name to suggest :D

File: cpp/src/copying/gather.cu [hunk 2]
```diff
@@ -25,16 +26,14 @@ struct dispatch_map_type {
...
-    bool check_bounds,
-    bool ignore_out_of_bounds,
-    bool allow_negative_indices         = false,
+    bounds check_bounds,
+    out_of_bounds oob,
+    negative_indices neg_indices        = negative_indices::NOT_ALLOWED,
     rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource(),
     cudaStream_t stream                 = 0)
   {
-    std::unique_ptr<table> destination_table;
```
Comment (vuule) [inline cpp/src/copying/gather.cu:34]: :+1:

File: cpp/include/cudf/detail/gather.hpp [hunk 3]
```diff
@@ -10,6 +10,13 @@
 namespace cudf {
 namespace experimental {
 namespace detail {
+
+// clang-format off
```
[the remark this answers is not in the document]
Comment (codereport) [reply cpp/include/cudf/detail/gather.hpp:14]: I like the ` : bool { `s aligned.

File: cpp/src/copying/gather.cu (hunk 2 above)
Comment (codereport) [reply cpp/src/copying/gather.cu:34]: :)

File: cpp/include/cudf/detail/gather.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (harrism) [reply cpp/include/cudf/detail/gather.hpp:14]: Wow, that's really worth two question-raising comment lines?

[the remark this answers is not in the document]
Comment (harrism) [reply cpp/include/cudf/detail/gather.hpp:14]: Doesn't clang-format have a setting for this?

File: cpp/include/cudf/detail/gather.hpp (hunk 1 above)
Comment (harrism) [reply cpp/include/cudf/detail/gather.hpp:52]: This is what I hate about using enum class for bool. It makes ridiculous parameter definitions that repeat the same word 3 times and cause you to use annoying abbreviations (oob) or unnecessary underscores.

I actually find these two parameters very confusing. What is the difference between "check_bounds" and "out of bounds"?  Do we need two separate parameters?  How can you not ignore out of bounds indices if you are not checking bounds? WTF?

Not saying this is less clear than the original, but I am saying this needs more care.

File: cpp/src/copying/gather.cu [hunk 4]
```diff
@@ -43,50 +42,39 @@ struct dispatch_map_type {
...
-                                               index_converter<map_type>{source_table.num_rows()}),
-               ignore_out_of_bounds,
-               mr,
-               stream);
+    if (neg_indices == negative_indices::ALLOW) {
+      auto idx_converter = index_converter<map_type>{source_table.num_rows()};
+      return gather(source_table,
+                    thrust::make_transform_iterator(gather_map.begin<map_type>(), idx_converter),
+                    thrust::make_transform_iterator(gather_map.end<map_type>(), idx_converter),
+                    oob == out_of_bounds::IGNORE,
```
Comment (harrism) [inline cpp/src/copying/gather.cu:50]: Why not fix the bool parameter in the gather API being called here?

File: cpp/include/cudf/detail/gather.hpp (hunk 1 above)
Comment (vuule) [reply cpp/include/cudf/detail/gather.hpp:52]: > It makes ridiculous parameter definitions that repeat the same word 3 times and cause you to use annoying abbreviations (oob) or unnecessary underscores.

True, but it makes the calls readable, which is arguably more important (single declaration, potentially many calls).

File: cpp/include/cudf/detail/gather.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (codereport) [reply cpp/include/cudf/detail/gather.hpp:14]: I can remove it.

File: cpp/include/cudf/detail/gather.hpp (hunk 1 above)
Comment (codereport) [reply cpp/include/cudf/detail/gather.hpp:50]: > I actually find these two parameters very confusing. What is the difference between "check_bounds" and "out of bounds"? Do we need two separate parameters? How can you not ignore out of bounds indices if you are not checking bounds? WTF?

I find these parameters confusing as well but I considered it out of scope for this refactoring. I didn't design the original API. These boolean parameters were already a part of the interface before the scoped enumerators were introduced.

File: cpp/src/copying/gather.cu (hunk 4 above)
Comment (codereport) [reply cpp/src/copying/gather.cu:50]: I thought about that but that API is not in the `detail` namespace and will break clients code. If we are OK with that I can definitely change that API as well.

File: cpp/include/cudf/detail/gather.hpp [hunk 5]
```diff
@@ -11,8 +11,8 @@ namespace cudf {
 namespace experimental {
 namespace detail {
 
-enum class bounds : bool { CHECK, NO_CHECK };
-enum class out_of_bounds : bool { IGNORE, DONT_IGNORE };
+enum class out_of_bounds_policy : int8_t { FAIL, NULLIFY, IGNORE };
```
Comment (harrism) [inline cpp/include/cudf/detail/gather.hpp:14]: This is great. But now not calling the other one `negative_indices_policy` seems strange.

Comment (codereport) [reply cpp/include/cudf/detail/gather.hpp:14]: Will update :)

Comment (harrism) [reply cpp/include/cudf/detail/gather.hpp:14]: Or 

```
enum class index_policy : bool { ALLOW_NEGATIVE, DISALLOW_NEGATIVE };
```

If someone adds another index policy that has nothing to do with negative indices later, we can change it then?

Comment (codereport) [reply cpp/include/cudf/detail/gather.hpp:14]: Done

File: cpp/src/copying/gather.cu [hunk 6]
```diff
@@ -115,7 +101,12 @@ std::unique_ptr<table> gather(table_view const& source_table,
                               rmm::mr::device_memory_resource* mr)
```
Comment (karthikeyann) [inline cpp/src/copying/gather.cu:101]: is `check_bounds`  okay to be `bool`?

Comment (codereport) [reply cpp/src/copying/gather.cu:101]: I talked with @harrism and seeing as it is a single `bool` (vs 3 back-to-back) and it is a non-`detail` `namespace` API, he agreed we should leave it.

File: cpp/include/cudf/detail/gather.hpp [hunk 7]
```diff
@@ -28,10 +33,10 @@ namespace detail {
  * @param[in] source_table The input columns whose rows will be gathered
  * @param[in] gather_map View into a non-nullable column of integral indices that maps the
  * rows in the source columns to rows in the destination columns.
- * @param[in] check_bounds Optionally perform bounds checking on the values
- * of `gather_map` and throw an error if any of its values are out of bounds.
- * @param[in] ignore_out_of_bounds Ignore values in `gather_map` that are
- * out of bounds. Currently incompatible with `allow_negative_indices`,
+ * @param[in] out_of_bounds_policy Specifies how to treat out of bounds. FAIL means checking if
```
[the remark this answers is not in the document]
Comment (codereport) [reply cpp/include/cudf/detail/gather.hpp]: Done

File: cpp/include/cudf/detail/gather.hpp [hunk 8]
```diff
@@ -28,10 +33,10 @@ namespace detail {
...
+ * the values of `gather_map` and throwing an error if any of its values are out of bounds.
+ * NULLIFY means to nullify indices that are outside the bounds. IGNORE means to ignore values in
+ * `gather_map` that are out of bounds. Currently incompatible with `allow_negative_indices`,
  * i.e., setting both to `true` is undefined.
```
[the remark this answers is not in the document]
Comment (codereport) [reply cpp/include/cudf/detail/gather.hpp]: Done.

File: cpp/include/cudf/detail/gather.hpp [hunk 9]
```diff
@@ -27,22 +32,21 @@ namespace detail {
...
- * @param[in] ignore_out_of_bounds Ignore values in `gather_map` that are
- * out of bounds. Currently incompatible with `allow_negative_indices`,
- * i.e., setting both to `true` is undefined.
- * @param[in] allow_negative_indices Interpret each negative index `i` in the
+ * @param[in] out_of_bounds_policy Specifies how to treat out of bounds indices. FAIL means checking
+ * if the values of `gather_map` and throwing an error if any of its values are out of bounds.
+ * NULLIFY means to nullify indices that are outside the bounds. IGNORE means to ignore values in
+ * `gather_map` that are out of bounds. The IGNORE option is currently incompatible when
+ * `negative_indices_policy` is set to ALLOW.
+ * @param[in] negative_indices_policy Interpret each negative index `i` in the
```
Comment (harrism) [inline cpp/include/cudf/detail/gather.hpp:40]: Trying to clean this up. Not sure about the correct use of "values" vs. "indices" in the below. Please review.

```suggestion
 * @param[in] out_of_bounds_policy How to treat out of bounds indices. FAIL: check `gather_map`
 * values and throw an exception if any are out of bounds. `NULLIFY` means to nullify output values
 * corresponding to out-of-bounds gather_map values. `IGNORE` means to ignore values in
 * `gather_map` that are out of bounds. `IGNORE` is incompatible with `negative_indices_policy == ALLOW`.
 * @param[in] negative_indices_policy Interpret each negative index `i` in the
```

File: cpp/include/cudf/detail/gather.hpp [hunk 10]
```diff
@@ -10,6 +10,11 @@
 namespace cudf {
 namespace experimental {
 namespace detail {
+
+enum class out_of_bounds_policy : int8_t { FAIL, NULLIFY, IGNORE };
+
+enum class negative_indices_policy : bool { ALLOW, NOT_ALLOWED };
```
[the remark this answers is not in the document]
Comment (codereport) [reply cpp/include/cudf/detail/gather.hpp]: Done.

File: cpp/src/copying/gather.cu [hunk 11]
```diff
@@ -23,16 +24,14 @@ struct dispatch_map_type {
...
+    negative_index_policy neg_indices,
     rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource(),
     cudaStream_t stream                 = 0)
   {
-    std::unique_ptr<table> destination_table;
-
-    if (check_bounds) {
-      cudf::size_type begin = (allow_negative_indices) ? -source_table.num_rows() : 0;
+    if (bounds == out_of_bounds_policy::FAIL) {
+      cudf::size_type begin =
```
Comment (karthikeyann) [inline cpp/src/copying/gather.cu:33]: which is  preferable here `cudf::size_type begin =` or `size_type begin = ` considering it's defined inside `namespace cudf { namespace detail {`

**Comment by codereport** [comment]
rerun tests

File: cpp/src/copying/gather.cu (hunk 11 above)
Comment (harrism) [reply cpp/src/copying/gather.cu:33]: I don't think it matters.

**Pull request merged**